audio-worklet/noisegate-test.html

```diff
@@ -0,0 +1,450 @@
+        let gain = new GainNode(context, {gain : 1});
+        let constantSource = new ConstantSourceNode(context, {gain: 1});
+        constantSource.connect(gain).connect(noisegate.input);
+        noisegate.output.connect(context.destination);
+
+        // The post-threshold gain should make samples less than the threshold.
+        // It must be greater than 0; otherwise samples will immediately
+        // be 0 after the threshold, and there will be no inspectable delay.
+        const beneathThresholdGain = 0.1;
+        gain.gain.setValueAtTime(beneathThresholdGain, timeToThreshold);
+        
+        constantSource.start();
+
+        // The noise gate output should contain four phases:
+        //   beforeOpenPhase (all 0) with a length determined by timeConstant.
+        //   openPhase (all 1) which ends at timeToThreshold.
+        //   betweenOpenAndClosePhase (all beneathThresholdGain) with a length
+        //       determined by timeConstant.
+        //   closePhase (all 0) for the remainder of the channel.
+        context.startRendering().then((buffer) => {
+          let channel = buffer.getChannelData(0);
+          
+          // With a constant source, the number of steps it takes for the
+          // envelope to open and close is a function of the time constant
+          // relative to the threshold.
+          const numOpeningFrames =
+              toIndex_(getTimeToTarget_(0, 1, timeConstant, linearThreshold),
+                       sampleRate);
+          
+          const numClosingFrames =
+              toIndex_(getTimeToTarget_(1, Math.pow(beneathThresholdGain, 2),
+                                        timeConstant, linearThreshold),
+                       sampleRate);
+
+          const openIndex = numOpeningFrames + scriptProcessorLatency;
+          
+          const sourceBeneathThresholdIndex =
+              testFrames * timeToThreshold + scriptProcessorLatency + 1;
+
+          let beforeOpenPhase =
+              channel.slice(scriptProcessorLatency, openIndex);
+          
+          let openPhase = channel.slice(openIndex, sourceBeneathThresholdIndex);
+
+          let betweenOpenAndClosePhase =
+              channel.slice(sourceBeneathThresholdIndex,
+                            sourceBeneathThresholdIndex + numClosingFrames);
+          
+          let closePhase = channel.slice(
+              sourceBeneathThresholdIndex + numClosingFrames);
+          
+          should(beforeOpenPhase, 'The samples before the gate opens')
+              .beConstantValueOf(0);
+          
+          should(openPhase, 'The open phase').beConstantValueOf(1);
+          
+          should(betweenOpenAndClosePhase, 'The samples before the gate closes')
+              .beConstantValueOf(Math.fround(beneathThresholdGain));
+
+          should(closePhase, 'The closed phase').beConstantValueOf(0);
+          
+          should(toTime_(numOpeningFrames, sampleRate), 'The opening delay')
```

Comment:
Sorry, I think I made a mistake here.  You compute `numOpeningFrames` based on a formula.  This value should be exactly the value you expect.  The threshold should be exactly zero.  But that makes the test meaningless.

What you should be doing is based on the output of the gate, find the frame where the gate actually does open.  Compare that frame number to `numOpeningFrames`.  These should not be off by more than 1.

But you basically have equivalent tests (I think) with the beforeOpenPhase and openPhase:  It's exactly 0 in the first part and exactly 1 in the second.  Thus, this test is not useful. I'd remove it.